File: src/main/java/hu/gerviba/webschop/service/CircleService.java
package hu.gerviba.webschop.service;

import java.util.List;

import javax.transaction.Transactional;

import hu.gerviba.webschop.dao.CircleEntityInfo;
import hu.gerviba.webschop.dao.OpeningRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import hu.gerviba.webschop.dao.CircleRepository;
import hu.gerviba.webschop.model.CircleEntity;

import static java.util.stream.Collectors.toList;

@Service
@Transactional
public class CircleService {

    @Autowired
    CircleRepository repo;

    @Autowired
    OpeningRepository openings;

    public Page<CircleEntity> findAll(Pageable pageable) {
        return repo.findAll(pageable);
    }
    
    public List<CircleEntity> findAll() {
        return repo.findAll();
    }

    public List<CircleEntity> findAllForMenu() {
        return repo.findAllByVisibleTrueOrderByHomePageOrder();
    }

    public List<CircleEntityInfo> findAllForInfo() {
        return repo.findAllByVisibleTrueOrderByHomePageOrder().stream()
                .map(CircleEntityInfo::new)
                .peek(info -> info.setNextOpening(openings
                        .findFirstByCircle_IdAndDateEndGreaterThanOrderByDateStart(
                                info.getCircleEntity().getId(), System.currentTimeMillis())
                        .orElse(null)))
                .collect(toList());
    }

    public void save(CircleEntity circleEntity) {
        repo.save(circleEntity);
    }
    
    public CircleEntity getOne(Long id) {
        return repo.getOne(id);
    }

    public void delete(CircleEntity ce) {
        repo.delete(ce);
    }

    public CircleEntity findByAlias(String alias) {
        return repo.findAllByAlias(alias).get(0);
    }
    
}

File: src/main/kotlin/hu/kirdev/schpincer/service/CircleService.kt
package hu.kirdev.schpincer.service

import hu.kirdev.schpincer.dao.CircleRepository
import hu.kirdev.schpincer.dao.OpeningRepository
import hu.kirdev.schpincer.dto.CircleEntityInfoDto
import hu.kirdev.schpincer.dto.RoleEntryDto
import hu.kirdev.schpincer.model.CircleEntity
import hu.kirdev.schpincer.web.sha256
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
open class CircleService {

    @Autowired
    private lateinit var repo: CircleRepository

    @Autowired
    private lateinit var openings: OpeningRepository

    @Transactional(readOnly = true)
    open fun findAll(pageable: Pageable): Page<CircleEntity> {
        return repo.findAll(pageable)
    }

    @Transactional(readOnly = true)
    open fun findAll(): List<CircleEntity> {
        return repo.findAll()
    }

    @Transactional(readOnly = true)
    open fun findAllForMenu(): List<CircleEntity> {
        return repo.findAllByVisibleTrueOrderByHomePageOrder()
    }

    @Transactional(readOnly = false)
    open fun save(circleEntity: CircleEntity) {
        repo.save(circleEntity)
    }

    private fun internalIncreaseInterestAndSave(circleEntity: CircleEntity): Long {
        val newInterest = circleEntity.interestCounter;
        repo.save(circleEntity)
        return newInterest
    }

    @Transactional(readOnly = false)
    open fun increaseInterest(id: Long): Long? {
        val circleEntity = getOne(id)
        if (circleEntity == null) {
            return null // TODO throw error?
        }
        return internalIncreaseInterestAndSave(circleEntity)
    }

    @Transactional(readOnly = false)
    open fun increaseInterestByAlias(alias: String): Long? {
        val circleEntity = findByAlias(alias)
        if (circleEntity == null) {
            return null // TODO throw error?
        }
        return internalIncreaseInterestAndSave(circleEntity)
    }

    @Transactional(readOnly = true)
    open fun getOne(id: Long): CircleEntity? {
        return repo.getOne(id)
    }

    @Transactional(readOnly = false)
    open fun delete(ce: CircleEntity) {
        repo.delete(ce)
    }

    @Transactional(readOnly = true)
    open fun findByAlias(alias: String): CircleEntity {
        return repo.findAllByAlias(alias)[0]
    }

    @Transactional(readOnly = true)
    open fun findByVirGroupId(id: Long): CircleEntity? {
        return repo.findOneByVirGroupId(id)
    }

    @Transactional(readOnly = true)
    open fun findAllForInfo(): List<CircleEntityInfoDto> {
        return repo.findAllByVisibleTrueOrderByHomePageOrder().asSequence()
                .map { CircleEntityInfoDto(it) }
                .onEach {
                    it.nextOpening = openings
                            .findFirstByCircle_IdAndDateEndGreaterThanOrderByDateStart(
                                    it.circleEntity.id, System.currentTimeMillis())
                            .orElse(null)
                }
                .toList()
    }

}